### ISSUE TITLE
CUDA outofmemory

### ISSUE THREAD

**Issue by zr526799544** [fanq15/FewX#25]
I do use my own custom coco dataset to train  FSOD by 4 gpus (GTX 2080 ti). cost 2315 MB per GPU when trainning the model and works. but when testing It always  CUDA out of memory whatever I do to fix the test image size. would you like to give me some advice?

**Comment by zr526799544** [comment]
CUDA out of memory. Tried to allocate 3.16 GiB (GPU 0; 10.76 GiB total capacity; 5.54 GiB already allocated; 2.72 GiB free; 7.15 GiB reserved in total by PyTorch)

**Comment by zr526799544** [comment]
I find it seems that I do wrong .datasets/generate_support_data.sh because when init the model It load 500+ support images. could you tell me what exactly generate_support_data.sh  do ？

**Comment by zr526799544** [comment]
I do not find the code for sample 10 suport images from coco. It seems that you code just do dump all the final_split_non_voc_instances_train2017.json which image size > 32 * 32.

**Comment by fanq15** [comment]
How many categories in your testing?

**Comment by zr526799544** [comment]
> How many categories in your testing?

|  category  | #instances   | category   | #instances   | category   | #instances   |
|:----------:|:-------------|:-----------|:-------------|:-----------|:-------------|
|     1      | 386          | 2          | 518          | 3          | 453          |
|     4      | 387          | 5          | 36           | 6          | 325          |
|     7      | 190          |            |              |            |              |
|   total    | 2295         |            |              |            |              |

**Comment by zr526799544** [comment]
> How many categories in your testing?

both train and test

**Comment by zr526799544** [comment]
why there is a code in fewx/modeling/fsod/fsod_roi_heads.py  155 assert len(proposal_boxes[0]) == 2000. why is 2000 ? how to calculate it?

**Comment by Fly-dream12** [comment]
I also encounter this problem. @fanq15

**Comment by fanq15** [comment]
> why there is a code in fewx/modeling/fsod/fsod_roi_heads.py 155 assert len(proposal_boxes[0]) == 2000. why is 2000 ? how to calculate it?

It is a hard coding for COCO dataset whose novel test set has 20 categories and each category has 100 proposals.

**Comment by fanq15** [comment]
> why there is a code in fewx/modeling/fsod/fsod_roi_heads.py 155 assert len(proposal_boxes[0]) == 2000. why is 2000 ? how to calculate it?

You should change the number according to your test category number.

**Comment by zr526799544** [comment]
> > why there is a code in fewx/modeling/fsod/fsod_roi_heads.py 155 assert len(proposal_boxes[0]) == 2000. why is 2000 ? how to calculate it?
> 
> You should change the number according to your test category number.

Yes........I did it.........but the test results not good.......I delete the code which will delete image that size < 32 * 32. because the category 3 and 4 is small target... what can I do to optimize it ？

**Comment by Fly-dream12** [comment]
How did you fix the cuda OOM problem, can you give some suggestions? When init the model,  it load 210 support images for a specific class, and it went OOM. Could you help? @zr526799544

**Comment by zr526799544** [comment]
> How did you fix the cuda OOM problem, can you give some suggestions? When init the model, it load 210 support images for a specific class, and it went OOM. Could you help? @zr526799544

you need to fix the data prepocess and just let you annfile has 10 * your categories support ann. that will reduce you suppor images

**Comment by Fly-dream12** [comment]
Do you mean final_split_{}_instances_trainval2017.json? @zr526799544

**Comment by zr526799544** [comment]
> Do you mean final_split_{}_instances_trainval2017.json? @zr526799544

Actually you need to fix train_support_df.pkl and 10_shot_support_df.pkl and because these pkl from  final_split_{}_instances_trainval2017.json so you can fix  final_split_{}_instances_trainval2017.json

**Comment by Fly-dream12** [comment]
I wonder what should be the proposal number in fewx/modeling/fsod/fsod_roi_heads.py assert len(proposal_boxes[0]) == 2000
My novel dataset is 3 class, however, here the length turns to be 100. Could you leave your contact information? @zr526799544

**Comment by zr526799544** [comment]
> I wonder what should be the proposal number in fewx/modeling/fsod/fsod_roi_heads.py assert len(proposal_boxes[0]) == 2000
> My novel dataset is 3 class, however, here the length turns to be 100. Could you leave your contact information? @zr526799544

I just delete the assert code.

**Comment by zr526799544** [comment]
> I wonder what should be the proposal number in fewx/modeling/fsod/fsod_roi_heads.py assert len(proposal_boxes[0]) == 2000
> My novel dataset is 3 class, however, here the length turns to be 100. Could you leave your contact information? @zr526799544

if you done it. would you like to tell me how it performance ?

**Comment by Fly-dream12** [comment]
how to contact you via wechat? @zr526799544

**Comment by xiaodongdreams** [comment]
> how to contact you via wechat? @zr526799544
Hello, Can I contact you via Wechat? I also encounter with the OOM Problem, and I find it hard to understand what to do to reduce the number of support images from @zr526799544 explanation. Would you please be more specific? Thanks.

**Comment by zr526799544** [comment]
> > how to contact you via wechat? @zr526799544
> > Hello, Can I contact you via Wechat? I also encounter with the OOM Problem, and I find it hard to understand what to do to reduce the number of support images from @zr526799544 explanation. Would you please be more specific? Thanks.

you can contact  me via mail . zr526799544@outlook.com

**Comment by fanq15** [comment]
> I wonder what should be the proposal number in fewx/modeling/fsod/fsod_roi_heads.py assert len(proposal_boxes[0]) == 2000
> My novel dataset is 3 class, however, here the length turns to be 100. Could you leave your contact information? @zr526799544

You can delete the assert, or change it to `100 * your class number`.

**Comment by Johnccl** [comment]
Hello, thanks for your good work. When I test a custom dataset with 50-way 5-shot on 2080TI. it always CUDA out of memory. would you like to give me some advice? @fanq15

**Comment by xiaodongdreams** [comment]
> > Do you mean final_split_{}_instances_trainval2017.json? @zr526799544
> 
> Actually you need to fix train_support_df.pkl and 10_shot_support_df.pkl and because these pkl from final_split_{}_instances_trainval2017.json so you can fix final_split_{}_instances_trainval2017.json

I have encountered with the same error, too. I tried the solution above, but it didn't help. Could you give us some advice? @zr526799544 @Fly-dream12

**Comment by xiaodongdreams** [comment]
> Do you mean final_split_{}_instances_trainval2017.json? @zr526799544 

Have you manage to overcome the OOM Problem?Could be so kind to give me some advice?

**Comment by zr526799544** [comment]
> > Do you mean final_split_{}_instances_trainval2017.json? @zr526799544
> 
> Have you manage to overcome the OOM Problem?Could be so kind to give me some advice?

you can show me the problem via email. zr526799544@outlook.com

**Comment by Zhang1Sheng** [comment]
> > I wonder what should be the proposal number in fewx/modeling/fsod/fsod_roi_heads.py assert len(proposal_boxes[0]) == 2000
> > My novel dataset is 3 class, however, here the length turns to be 100. Could you leave your contact information? @zr526799544
> 
> You can delete the assert, or change it to `100 * your class number`.

Although I changed ‘assert len(proposal_boxes[0]) == 2000’ to assert ”len(proposal_boxes[0]) == 100 * my class number”, it only predicted around 20 Bboxes per class, which is far less than 100. Shouldn't I generate 100? How to solve this problem?

**Comment by xiaodongdreams** [comment]
> > > Do you mean final_split_{}_instances_trainval2017.json? @zr526799544
> 
> > 
> 
> > Have you manage to overcome the OOM Problem?Could be so kind to give me some advice?
> 
> 
> 
> you can show me the problem via email. zr526799544@outlook.com

Could you kindly check your e-mail please? I heard no response from you for a long time. Thank you very much.

**Comment by Oiolosse-mr** [comment]
> > > > Do you mean final_split_{}_instances_trainval2017.json? @zr526799544
> > 
> > 
> > > 
> > 
> > 
> > > Have you manage to overcome the OOM Problem?Could be so kind to give me some advice?
> > 
> > 
> > you can show me the problem via email. [zr526799544@outlook.com](mailto:zr526799544@outlook.com)
> 
> Could you kindly check your e-mail please? I heard no response from you for a long time. Thank you very much.

Hi, I also met the OOM problem and I wonder if you have manage to slove the problem? Could you please give me some help on how to overcome this? Thank you.
